transformer_engine/common/common.h

```diff
@@ -281,6 +281,185 @@ struct Tensor {
+  }
+
+  void *const *get_dptr_list() const {
+    NVTE_CHECK(has_dptr_list(), "Data pointers must be set");
+    return dptr_list.data();
+  }
+
+  void *get_base_dptr() const {
+    NVTE_CHECK(base_dptr != nullptr, "Base pointer must be set");
+    NVTE_CHECK(contiguous, "Base pointer can be used only when data is contiguous");
+    return base_dptr;
+  }
+
+  bool has_data() const { return base_dptr != nullptr || !dptr_list.empty(); }
+
+  void clear() {
+    dptr_list.clear();
+    base_dptr = nullptr;
+    dtype = DType::kFloat32;
+    num_tensors = 0;
+    contiguous = false;
+    sum_first_dims = 0;
+    sum_second_dims = 0;
+  }
+};
+
+struct GroupedTensor {
+ public:
+  /*
+  Grouped tensor is a collection of tensors with different shapes but the same dtype and scaling mode
+  1. 1D linear layout for all fields: tensor1(dim1, dim2) | tensor2(dim1, dim2) | ... | tensorN(dim1, dim2)
+  2. All the SimpleGroupedTensor fields should store device pointer(s)
+  3. All the kernels that use GroupedTensor should check the contiguous flag and act accordingly.
+     In the case if only either contiguous or non-contiguous is supported, make sure to error out for the other case.
+  */
+
+  SimpleGroupedTensor data;
+  SimpleGroupedTensor columnwise_data;
+  SimpleGroupedTensor scale_inv;
+  SimpleGroupedTensor columnwise_scale_inv;
+  SimpleGroupedTensor amax;
+  SimpleGroupedTensor columnwise_amax;
+  SimpleGroupedTensor scale;  // for FP8-DS only
+
+  // [TODO] Discuss whether the first_dims and second_dims should be according to layout N
+  // Shape information: first_dims[i] and second_dims[i] define the shape of the i-th tensor
+  // For 2D tensors: shape[i] = (first_dims[i], second_dims[i])
+  SimpleTensor first_dims;   // Device pointer to size_t array of length num_tensors
+  SimpleTensor second_dims;  // Device pointer to size_t array of length num_tensors
+
+  // Cumulative sizes for indexing into contiguous layout
+  // cumulative_tensor_sizes[i] = sum of numel for tensors 0..i-1
+  SimpleTensor cumulative_tensor_sizes;  // Device pointer to size_t array of length num_tensors+1
+
+  NVTEScalingMode scaling_mode;
+  NVTEGroupedTensor nvte_tensor;
+
+  GroupedTensor(NVTEScalingMode scaling_mode)
+      : data(),
+        columnwise_data(),
+        scale_inv(),
+        columnwise_scale_inv(),
+        amax(),
+        columnwise_amax(),
+        scale(),
+        first_dims(nullptr, {0}, DType::kInt64),
+        second_dims(nullptr, {0}, DType::kInt64),
+        cumulative_tensor_sizes(nullptr, {0}, DType::kInt64),
+        scaling_mode(scaling_mode),
+        nvte_tensor(0) {}
+
+  explicit operator NVTEGroupedTensor() const noexcept { return nvte_tensor; }
+
+  bool has_data() const noexcept { return data.has_data(); }
+  bool has_columnwise_data() const noexcept { return columnwise_data.has_data(); }
+  bool has_cumulative_tensor_sizes() const noexcept {
+    return cumulative_tensor_sizes.dptr != nullptr;
+  }
```

Comment:
Without those the tensor would be kind of unusable though - it should always have them.